ci: update publish workflow to work on tag push

joanise opened the pull request:
<!-- PR template: please provide enough information to guide your reviewers.
Please read Contributing.md before submitting a PR.  -->

### PR Goal? <!-- Explain the main objective of this PR. -->

Change the publication workflow to work on tag push instead of PR to release

### Feedback sought? <!-- What should reviewers focus on in particular? -->

- how to test this properly
- careful review since it's going to be our publication process

### Priority? <!-- How soon would you like this PR reviewed, does it block other work? -->

high

### Tests added? <!-- Make sure your PR includes automated tests for your changes. -->

n/a

### How to test? <!-- Explain how reviewers should test this PR. -->

Good question,  still need to figure this out!

### Confidence? <!-- How confident are you that these changes are ready to merge? -->

medium

### Version change? <!-- Do you think this PR should trigger a Major (Breaking Change)/Minor (New Feature)/patch (refactor/bug fix) version change? -->

no

<!-- Add any other relevant information here -->

joanise committed:
ci: update publish workflow to work on tag push

joanise committed:
style: break up multiple commands on multiple lines

roedoejet commented on .github/workflows/release.yml:
```diff
@@ -44,33 +36,17 @@ jobs:
+      - name: Update CHANGELOG
+        id: changelog
+        uses: requarks/changelog-action@v1
         with:
```
it might be nice to have a check here that determines whether npm has the version available (i.e. whether we forgot to bump). `npm view @readalongs/web-component version` should display the most recent version, and that should be less than the current defined version in the code/tag

roedoejet replied:
if it doesn't, we shouldn't create a github release, and this should fail.

joanise replied:
Will `npm publish` fail if you try to republish the same version again?

joanise replied:
But good point, we should make sure the GH release doesn't happen in this case.

joanise replied:
Indeed, we should already be safe: https://docs.npmjs.com/cli/v8/commands/npm-publish says:
> The publish will fail if the package name and version combination already exists in the specified registry.

and once the first publish fails the entire workflow will abort.

roedoejet replied:
hmm I feel like if that's the case, it's new. I seem to remember things not aborting just because npm publish failed

roedoejet commented on README.md:
```diff
@@ -185,27 +185,30 @@ or
+NPM. Then, apply all the version bump and `npm install` steps above,
+and build both the web component and angular wrapper:
 
     npx nx build web-component
```
Don't we need to also run `npx nx bundle web-component`?

joanise replied:
That's just a few lines lower in the README.md. It's needed before running `npx nx build studio-web`, but since we don't publish studio-web it's not that relevant...

joanise replied:
- Updated the readme to say you need to do this before even creating the tag.
 - Updated the release workflow to fail early if the bundle is out of date.

roedoejet reviewed (changes requested):
Looks great @joanise - and super quickly done 👏 

I've made a request for something that checks that we bumped the version, and also a clarifying question for the manual publishing.

joanise commented:
Tested the workflow by creating a variant that ran on push to `test1.2.3` and did dry run operations.
See https://github.com/ReadAlongs/Studio-Web/actions/runs/8914132259/job/24481166770
It looks like it works fine.

joanise committed:
ci: make sure the bundle is up to date before publishing

dhdaines commented on .github/workflows/release.yml:
```diff
@@ -1,76 +1,60 @@
+          npx nx build web-component
+          npx nx bundle web-component
+          npx nx build ngx-web-component
+      - name: Check that the web-component bundle was up to date
```
We could put a check here to make sure the dependency in ngx-web-component is correct as well, since we're asking people to do it manually in the README.md..

dhdaines replied:
Added this in 2604892

dhdaines commented on README.md:
```diff
@@ -185,27 +185,31 @@ or
+- Update the version number for web-component and ngx-web-component in [packages/web-component/package.json](packages/web-component/package.json) and [packages/ngx-web-component/package.json](packages/ngx-web-component/package.json).
+- Nx does not manage dependencies among projects for us, so update the version in `peerDependencies` for `@readalongs/web-component` in [packages/ngx-web-component/package.json](packages/ngx-web-component/package.json).
+- Update `singleFileBundleVersion` in [packages/studio-web/package.json](packages/studio-web/package.json).
+- All the versions above should be exactly in sync!
```
As mentioned above... we should verify this in the release workflow and fail if not.  I will try adding some code to do this shortly.

dhdaines committed:
feat: add check for versions in sync

dhdaines reviewed (approved):
Okay, now I approve, with the change I added to verify the versions.  I tested it on the command-line locally, unsure how to test it in actual CI but it seems correct.

joanise committed:
ci: reject release if the tag fails any version validation

 - must be consistent in the repo and the tag
 - must be more recent than latest on NPM

This solution requires jq, but that's OK because every runner image on
GitHub Actions has jq pre-installed.

Some of the tests with jq are equivalent to, e.g.,
`npm view dist/packages/web-component version`
but doing them with `jq` allows us to apply them earlier, before we even
make the builds.

joanise commented:
Darn, @dhdaines we worked in parallel. Before leaving work yesterday, I had almost completed the same checks you did later in the evening. See https://github.com/ReadAlongs/Studio-Web/commit/decc1d72b2faf0168820fc9404bf953e5a51a652 
My checks are a little more complete, but use jq instead of npm. Do you have a preference between your solution and mine?

dhdaines commented:
> Darn, @dhdaines we worked in parallel. Before leaving work yesterday, I had almost completed the same checks you did later in the evening. See [decc1d7](https://github.com/ReadAlongs/Studio-Web/commit/decc1d72b2faf0168820fc9404bf953e5a51a652) My checks are a little more complete, but use jq instead of npm. Do you have a preference between your solution and mine?

Yours are more complete but I think it would be preferable to use `npm` since it's already there... should be pretty easy to just drop in the `npm` calls that I did.

joanise commented:
> Yours are more complete but I think it would be preferable to use `npm` since it's already there... should be pretty easy to just drop in the `npm` calls that I did.

`jq` is there by default on all the runner images on GitHub, so being there is not an issue. I like how `jq` can run earlier, before we even build, and thus fail early. But I like how your `npm` commands run on actual builds in `dist/`. (And, ah-ha! that's how to do it locally! I tried and failed to get `npm view` to dome something useful with `packages/web-component`.)

I don't see an `npm view` equivalent for `"$(jq -r .singleFileBundleVersion packages/studio-web/package.json)" != "$VERSION"`, though, since we don't build or publish that package, but we still want to check this.

dhdaines commented:
Ah, good to know ... so no problem to use `jq`, or `npm view` either for that matter :) your checks are more complete and probably more correct.

joanise commented:
> Ah, good to know ... so no problem to use `jq`, or `npm view` either for that matter :) your checks are more complete and probably more correct.

Thanks for doing yours, though, and sorry for not flagging that I was writing them yesterday.
